File: demo_lazyEval_meanStd.R
if (!require("this.path", quietly = TRUE))
{
  install.packages("this.path")
}
library(this.path)

if (!requireNamespace("BiocManager", quietly = TRUE))
{
  install.packages("BiocManager")
}
if (!require("DelayedMatrixStats", quiety = TRUE))
{
 BiocManager::install("DelayedMatrixStats")
}
if (!require("HDF5Array", quiety = TRUE))
{
 BiocManager::install("HDF5Array")
}

library(HDF5Array)
library(DelayedMatrixStats)
library(DelayedArray)

# Load HDF5 file as a DelayedArray; the variable of interest is called data
workDir         <- this.path::this.dir()
inputFile       <- file.path(workDir, "samples", "NIfTI", "concatedData_73_uncompressed.mat")
dataset_delayed <- HDF5Array(inputFile, name = "data")

# Calculate mean, per column (i.e., voxel-wise)
voxelAverages <- colMeans2(dataset_delayed)

# Calculate standard deviation per column (i.e., voxel-wise)
voxelStd <- colSds(dataset_delayed)

# Show results
cat(paste0("Mean of the data is: ",               round(voxelAverages, 4), "\n"))
cat(paste0("Standard deviation of the data is: ", round(voxelStd, 4), "\n"))

File: demo_lazyEval_meanStd.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu May 28 18:06:26 2026

@author: praveshp
"""

import h5py
import dask
import dask.array
import dask.dataframe
import os

# Which file to work on; variable is called data
workDir = os.path.dirname(os.path.realpath(__file__))
toWork  = os.path.join(workDir, "samples", "NIfTI", "concatedData_73_uncompressed.mat")

# Keep file open while reading
with h5py.File(toWork, mode='r') as f:
    delayed_df = dask.array.from_array(f['/data'], chunks='auto')

    # Define our lazy mean and standard functions
    lazy_mean = delayed_df.mean(axis=1)
    lazy_std  = delayed_df.std(axis=1, ddof=1)

    # Get results
    [mean, std] = dask.compute(lazy_mean, lazy_std)


# Show the results
print(f"Mean of the data is: {mean}")
print(f"Standard deviation of the data is: {std}")

# Older solution:
# Open the file
# f = h5py.File(toWork, mode='r')
# [rows, cols] = f['/data'].shape
# value = dask.delayed(f['/data'])
# delayed_df = dask.array.from_delayed(value, (rows,cols), dtype=float)

# Define our lazy mean and standard functions
# lazy_mean = delayed_df.mean(axis=1)
# lazy_std  = delayed_df.std(axis=1, ddof=1)

# Get results
# [mean, std] = dask.compute(lazy_mean, lazy_std)

# Close the file
# f.close()